### PR TITLE
Quick fix for handling evidences from different collection tools

### PR DESCRIPTION
**Pull request by garuppel**
**Problem**
We have to somehow deal in the evaluation with evidences from different discoverers for the same security feature.  

So far, a rego policy is evaluated if the parent object of an security feature is available (e.g. MalwareProtection). But currently we nil it in the Clouditor since we cannot determine the corresponding security features (e.g. MalwareProtectionEnabled).

**Solution**
For a solution now, we only evaluate evidences with metrics if the corresponding fields (security features, not just parent) are not nil. For example, since the Clouditor is currently not able to fill the MalwareProtection, the evaluation for the metricIDs MalwareProtectionEnabled and MalwareProtectionOutput will not be evaluated. This is done by other tools in the MEDINA context.

For other security features other tools can not discover, we set it it to a default value (false, 0, "" ...).

**Alternatives**
We'll have to figure out how to deal with the problem later #659.

### CODE REVIEW

**Comment by lebogg** [comment]
The new approach is:
1) Cached metrics are tied to resource types AND the tool. This allows discovering an entire evidence by splitting the work to different discoverers (e.g. Clouditor and Wazuh in Medina). Such a split can avoid multiple assessment results for the same evidence.
2) For 1) to work, we now check applicability by checking if a security feature is set (not null). This allows a finer granularity for deciding which security features (and therefore metrics) are evaluated by the respective tool. Before the whole parent (e.g. atRestEncryption) was checked instead of a single security feature (atRestEncryption.enabled).

File: policies/bundles/OSLoggingEnabled/metric.rego [hunk 1]
```diff
@@ -6,12 +6,12 @@ default applicable = false
 
 default compliant = false
 
-OSLogging := input.oSLogging
+enabled := input.oSLogging.enabled
```
Comment (garuppel) [inline policies/bundles/OSLoggingEnabled/metric.rego:9]:
```suggestion
enabled := input.oSLogging.logging.enabled
```

File: policies/bundles/MalwareProtectionOutput/metric.rego [hunk 2]
```diff
@@ -6,12 +6,12 @@ default applicable = false
 
 default compliant = false
 
-mp := input.malwareProtection.applicationLogging
+output := input.malwareProtection.applicationLogging.loggingService
```
Comment (garuppel) [inline policies/bundles/MalwareProtectionOutput/metric.rego:9]:
```suggestion
output := input.malwareProtection.applicationLogging.logging.loggingService
```

File: policies/bundles/OSLoggingOutput/metric.rego [hunk 3]
```diff
@@ -8,12 +8,12 @@ default compliant = false
 
 metricConfiguration := data.target_value
 
-OSLogging := input.oSLogging
+output := input.oSLogging.loggingService
```
Comment (garuppel) [inline policies/bundles/OSLoggingOutput/metric.rego:11]:
```suggestion
output := input.oSLogging.logging.loggingService
```

File: policies/bundles/OSLoggingRetention/metric.rego [hunk 4]
```diff
@@ -6,12 +6,12 @@ default applicable = false
 
 default compliant = false
 
-OSLogging := input.oSLogging
+retentionPeriod := input.oSLogging.retentionPeriod
```
Comment (garuppel) [inline policies/bundles/OSLoggingRetention/metric.rego:9]:
```suggestion
retentionPeriod := input.oSLogging.logging.retentionPeriod
```

File: policies/bundles/BootLoggingEnabled/metric.rego [hunk 5]
```diff
@@ -6,12 +6,12 @@ default applicable = false
 
 default compliant = false
 
-bootLogging := input.bootLogging
+enabled := input.bootLogging.enabled
```
Comment (garuppel) [inline policies/bundles/BootLoggingEnabled/metric.rego:9]:
```suggestion
enabled := input.bootLogging.logging.enabled
```

File: policies/bundles/BootLoggingOutput/metric.rego [hunk 6]
```diff
@@ -8,12 +8,12 @@ default compliant = false
 
 metricConfiguration := data.target_value
 
-bootLogging := input.bootLogging
+output := input.bootLogging.loggingService
```
Comment (garuppel) [inline policies/bundles/BootLoggingOutput/metric.rego:11]:
```suggestion
output := input.bootLogging.logging.loggingService
```

File: policies/bundles/BootLoggingRetention/metric.rego [hunk 7]
```diff
@@ -6,12 +6,12 @@ default applicable = false
 
 default compliant = false
 
-bootLogging := input.bootLogging
+retentionPeriod := input.bootLogging.retentionPeriod
```
Comment (garuppel) [inline policies/bundles/BootLoggingRetention/metric.rego:9]:
```suggestion
retentionPeriod := input.bootLogging.logging.retentionPeriod
```

File: policies/bundles/SingleSignOnEnabled/metric.rego [hunk 8]
```diff
@@ -6,12 +6,12 @@ default applicable = false
 
 default compliant = false
 
-sso := input.singleSignOn
+enabled := input.singleSignOn.enabled
```
Comment (garuppel) [inline policies/bundles/SingleSignOnEnabled/metric.rego:9]: SingleSignOn.enabled does not yet exist in the Ontology.

File: policies/bundles/TransportEncryptionAlgorithm/metric.rego [hunk 9]
```diff
@@ -7,10 +7,12 @@ default compliant = false
 
 default applicable = false
 
+algorithm := endpoint.transportEncryption.algorithm
```
Comment (garuppel) [inline policies/bundles/TransportEncryptionAlgorithm/metric.rego:10]: Where comes the endpoint from? 
Shouldn't it be `input.httpEndpoint.transportEncryption.algorithm`?

Comment (garuppel) [inline policies/bundles/TransportEncryptionAlgorithm/metric.rego:10]:
```suggestion
algorithm := input.httpEndpoint.transportEncryption.algorithm
```

File: policies/policies.go [hunk 10]
```diff
@@ -39,7 +39,7 @@ var (
 	log = logrus.WithField("component", "policies")
 )
 
-type metricsResourceTypeCache struct {
+type metricsToolIDAndResourceTypeCache struct {
```
Comment (garuppel) [inline policies/policies.go:42]: I like that name :)

File: policies/policies.go [hunk 11]
```diff
@@ -66,8 +66,10 @@ type MetricsSource interface {
 	MetricImplementation(lang assessment.MetricImplementation_Language, metric string) (*assessment.MetricImplementation, error)
 }
 
-func createKey(types []string) (key string) {
-	key = strings.Join(types, "-")
+// createKey creates a key by concatenating toolID and all types
+func createKey(toolID string, types []string) (key string) {
+	// Merge toolID and types to one slice and concatenate all its elements
+	key = strings.Join(append(types, toolID), "-")
 	key = strings.ReplaceAll(key, " ", "")
```
Comment (garuppel) [inline policies/policies.go:73]: Does the key (types and/or toolIds) really have spaces?

File: service/discovery/azure/compute.go [hunk 12]
```diff
@@ -172,7 +172,13 @@ func (d *azureComputeDiscovery) handleVirtualMachines(vm *compute.VirtualMachine
...
 			RetentionPeriod: 0, // Currently, configuring the retention period for Managed Boot Diagnostics is not available. The logs will be overwritten after 1gb of space according to https://github.com/MicrosoftDocs/azure-docs/issues/69953
 		}},
-		OSLogging:    &voc.OSLogging{}, // TODO(garuppel): Add OSLogging
+		OSLogging: &voc.OSLogging{
+			Logging: &voc.Logging{
+				Enabled:         false,
+				RetentionPeriod: 0,
+				LoggingService:  []voc.ResourceID{},
+			},
+		}, // TODO(garuppel): Add OSLogging
```
Comment (garuppel) [inline service/discovery/azure/compute.go:181]:
```suggestion
		}, 
```

File: policies/bundles/AnomalyDetectionOutput/metric.rego [hunk 13]
```diff
@@ -6,12 +6,12 @@ default applicable = false
 
 default compliant = false
 
-ad := input.anomalyDetection
+output := input.anomalyDetection.output
```
Comment (garuppel) [inline policies/bundles/AnomalyDetectionOutput/metric.rego:9]:
```suggestion
output := input.anomalyDetection.applicationLogging.logging.loggingService
```

File: policies/bundles/AnonymousAuthentication/metric.rego [hunk 14]
```diff
@@ -6,12 +6,12 @@ default applicable = false
 
 default compliant = false
 
-cba := input.certificateBasedAuthentication
+enabled := input.certificateBasedAuthentication.enabled
```
Comment (garuppel) [inline policies/bundles/AnonymousAuthentication/metric.rego:9]: CertificateBasedAuthentication does not have a filed enabled yet
.

**Review by garuppel — COMMENTED** [review]
Some minor comments .

File: policies/bundles/MalwareProtectionEnabled/metric.rego [hunk 15]
```diff
@@ -6,12 +6,12 @@ default applicable = false
 
 default compliant = false
 
-mp := input.malwareProtection
+enabled := input.malwareProtection.enabled
```
Comment (garuppel) [inline policies/bundles/MalwareProtectionEnabled/metric.rego:9]:
```suggestion
enabled := input.malwareProtection.applicationLogging.enabled
```

File: policies/policies.go (hunk 11 above)
Comment (lebogg) [reply policies/policies.go:73]: Hopefully not. But it is not completely under our control.

File: policies/policies.go (hunk 10 above)
Comment (lebogg) [reply policies/policies.go:42]: I would prefer just metricsCache and write a comment whats the key of. I didn't want to change it though. Maybe I will later :D

File: policies/bundles/AnomalyDetectionOutput/metric.rego (hunk 13 above)
Comment (lebogg) [reply policies/bundles/AnomalyDetectionOutput/metric.rego]: I changed it (without ".logging." as discussed) but I'm not 100% sure if we set this right in the ontology. Anyway, we will figure it out when we will try to implement it.

File: policies/bundles/TransportEncryptionAlgorithm/metric.rego (hunk 9 above)
Comment (lebogg) [reply policies/bundles/TransportEncryptionAlgorithm/metric.rego:10]: From the import in L4

Comment (garuppel) [reply policies/bundles/TransportEncryptionAlgorithm/metric.rego:10]: so true...

**Pull request merged**